lib/rules/no-unused-class-component-methods.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * @fileoverview Prevent declaring unused methods of component class
+ * @author Paweł Nowak
+ */
+
+'use strict';
+
+const Components = require('../util/Components');
+const astUtil = require('../util/ast');
+const docsUrl = require('../util/docsUrl');
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+const internalMethods = [
+  'constructor',
+  'componentDidCatch',
+  'componentDidMount',
+  'componentDidUpdate',
+  'componentWillMount',
+  'componentWillReceiveProps',
+  'componentWillUnmount',
+  'componentWillUpdate',
+  'getSnapshotBeforeUpdate',
+  'render',
+  'shouldComponentUpdate',
+  'UNSAFE_componentWillMount',
+  'UNSAFE_componentWillReceiveProps',
+  'UNSAFE_componentWillUpdate'
+];
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Prevent declaring unused methods of component class',
+      category: 'Best Practices',
+      recommended: false,
+      url: docsUrl('no-unused-class-component-methods')
+    },
+    schema: [
+      {
+        type: 'object',
+        additionalProperties: false
+      }
+    ]
+  },
+
+  create: Components.detect((context, components, utils) => {
+    const isNotComponent = node => (
+      !utils.isES5Component(node) &&
+      !utils.isES6Component(node) &&
+      !utils.isCreateElement(node)
+    );
+    const filterAllMethods = node => {
+      const isMethod = node.type === 'MethodDefinition';
+      const isArrowFunction = (
+        node.type === 'ClassProperty' &&
+        node.value.type === 'ArrowFunctionExpression'
+      );
+      return isMethod || isArrowFunction;
+    };
+    const checkMethods = node => {
+      if (isNotComponent(node)) {
+        return;
+      }
+      const properties = astUtil.getComponentProperties(node);
+      let methods = properties
+        .filter(property => (
+          filterAllMethods(property) &&
+          !internalMethods.includes(astUtil.getPropertyName(property))
+        ));
+      const getThisExpressions = subnode => {
+        if (!methods.length) {
+          return;
+        }
+
+        switch (subnode.type) {
+          case 'ClassProperty':
+          case 'JSXAttribute':
+          case 'MethodDefinition':
+            getThisExpressions(subnode.value);
```

Comment:
Consider using the Object switch methodology rather than an actual switch, in my view, it's a lot faster than the switch. 

See this link for a reference. 
https://stackoverflow.com/questions/37730199/switch-vs-object-lookup-performance-since-jsperf-is-down

Reply:
Faster or not, it's always cleaner :-)